hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/DefaultS3ClientFactory.java

```diff
@@ -138,14 +157,21 @@ private <BuilderT extends S3BaseClientBuilder<BuilderT, ClientT>, ClientT> Build
       BuilderT builder, S3ClientCreationParameters parameters, Configuration conf, String bucket)
       throws IOException {
 
-    Region region = parameters.getRegion();
-    LOG.debug("Using region {}", region);
-
     URI endpoint = getS3Endpoint(parameters.getEndpoint(), conf);
 
+    Region region = null;
+
     if (endpoint != null) {
       builder.endpointOverride(endpoint);
-      LOG.debug("Using endpoint {}", endpoint);
+      region = getS3RegionFromEndpoint(parameters.getEndpoint());
+      LOG.debug("Using endpoint {} and region {}", endpoint, region);
+    }
+
+    if (region != null) {
+      builder.region(region);
+    } else {
+      builder.crossRegionAccessEnabled(true);
+      builder.region(getS3Region(conf));
```

Comment:
if a region is set in the config, it must take priority over anything the AWS SDK determines from the endpoint. without that we can never correct for sdk issues

Reply:
also, what about: env vars and sysprops? good or bad?